src/event-helper.js

```diff
@@ -152,18 +153,32 @@ export function loadPromise(eleOrWindow) {
     return Promise.resolve(eleOrWindow);
   }
   const loadingPromise = new Promise((resolve, reject) => {
+    const isMediaElement = isHTMLMediaElement(eleOrWindow);
     // Listen once since IE 5/6/7 fire the onload event continuously for
     // animated GIFs.
-    const {tagName} = eleOrWindow;
-    if (tagName === 'AUDIO' || tagName === 'VIDEO') {
-      unlistenLoad = listenOnce(eleOrWindow, 'loadstart', resolve);
+    if (isMediaElement) {
+      // The following event can be triggered by the media or one of its
+      // sources. Using capture is required as the media events do not bubble.
+      unlistenLoad =
+          listenOnce(eleOrWindow, 'loadedmetadata', resolve, {capture: true});
     } else {
       unlistenLoad = listenOnce(eleOrWindow, 'load', resolve);
     }
-    // For elements, unlisten on error (don't for Windows).
-    if (tagName) {
-      unlistenError = listenOnce(eleOrWindow, 'error', reject);
+    // Don't unlisten on error for Windows.
+    if (!eleOrWindow.tagName) {
+      return;
     }
+    let errorTarget = eleOrWindow;
+    // If the media element has no `src`, it will try to load the sources in
+    // document order. If the last source errors, then the media element
+    // loading errored.
+    if (isMediaElement && !eleOrWindow.hasAttribute('src')) {
+      errorTarget = eleOrWindow.querySelector('source:last-of-type');
```

Comment:
Can source elements be nested inside another element?

Reply:
I meant for `<audio>` and `<video>` specifically:

```html
<video width="320" height="240" controls>
  <div>
    <source src="movie.mp4" type="video/mp4">
  </div>
  <div>
    <source src="movie.ogg" type="video/ogg">
  </div>
  Your browser does not support the video tag.
</video>
```

It seems it has to be a direct child for it work, so we can switch to https://github.com/ampproject/amphtml/blob/9ae3a653ba00dc9a3099bfdd60ae550230ecf7b4/src/dom.js#L367-L381 to get a speed boost.

Reply:
Cool! Done